lib/linalg.dx

```diff
@@ -193,40 +277,32 @@ def lu {n} (a: n=>n=>Float) :
   --      lRef!i!j := (a.i.j - s) / (get uRef!j!j)
   --    for i:n. ()
 
-    -- Helper functions to index into triangular matrices
-    -- from indices that themselves need to be cast or projected
-    -- to the right type.
-    def unsafe_lmat {m} (i:n) (j:m) [Ix m] : Ref _ Float =
-      lower_tri_mat lRef i $ unsafe_cast j
-    def unsafe_umat {m} (i:n) (j:m) [Ix m, Subset (i..) m] : Ref _ Float =
-      upper_tri_mat uRef i $ from_just (project (i..) j)
-
     for j:n.
       for i:(..j).
-        i' = inject n i
         s = sum for k:(..i).
-          -- TODO: get transitivity of Subset working so we can do a single injection
-          k' = inject n (inject (..j) k)
-          ukj = get $ unsafe_umat k' j
-          lik = get $ unsafe_lmat i' k
+          (MkUpperTriIx k2 _) = transpose_lower_ix i k
+          (MkUpperTriIx k3 j3) = transpose_lower_ix j k2
```

Comment:
Kudos to you, I did not feel close to putting all this indexing into safe calls.  Although from a readability point of view its still hard on the brain - this double transpose really gets my noggin joggin.  However, I did originally copy this algorithm from a C version instead of deriving it from first principles, so maybe we've lost the unifying structure.  Perhaps if Conal Elliot rewrote this algorithm from scratch we would see its true unified nature and it would look great.

Reply:
I admit, it was something of a jigsaw puzzle to come up with these.  My goal was just to try and minimize the surface of trust, not necessarily to make the program readable :)